### PR TITLE
log download activity when using cloudstorage

### PR DESCRIPTION
**Pull request by chris48s**
Refs #269
Refs #254

This _should_ work, but I've not actually tested it using the cloudstorage backend yet

### CODE REVIEW

**Comment by chris48s** [comment]
I tried deploying this to UAT to test it with S3 and it.. doesn't work :angry:  The issue is:

With the file storage, `resource_download` returns a response, so we can do:

```py
resp = super(BaseController, self).resource_download(
    id, resource_id, filename=filename
)
self._log_download_activity(context, resource_id)
return resp
```

With the cloudstorage controller, `resource_download` issues a redirect, so if we try to do

```py
resp = super(BaseController, self).resource_download(
    id, resource_id, filename=filename
)
self._log_download_activity(context, resource_id)
return resp
```

`resource_download` issues a redirect and we never execute `self._log_download_activity(context, resource_id)`. We've already issued a response.

The things I can think of to do here are:

1. Log the download first when we're using cloudstorage, which is simpler but will probably generate some false-positives (as you correctly pointed out in https://github.com/okfn/ckanext-unhcr/pull/280#discussion_r412041624 )
2. Completely override `resource_download`, copy and paste the entire `resource_download` method from https://github.com/TkTech/ckanext-cloudstorage/blob/master/ckanext/cloudstorage/controller.py into our plugin and then we can call our `log_download_activity()` before `h.redirect_to()`.
3. Fork (or see if we can contribute to) `cloudstorage` to make this easier

@amercader - what are your thoughts on which way to go on this, or do you have another smarter idea?

**Comment by chris48s** [comment]
I've not tried it yet, but over the weekend I thought of another option, which is that we could probably use `mock.patch()` to inject our own `h.redirect_to()` and log the download activity like that.

**Comment by amercader** [comment]
@chris48s I think I'd go for the simple approach and choose option 1. By the point we log the download we'd have checked if the resource exists and the authorization which would be the most common user errors. Any further exceptions would come from cloudstorage itself, and we hopefully will pick these and fix them so I'd say the danger of false positives is low.

CKAN 2.9 might introduce signals (https://github.com/ckan/ckan/pull/5359) which will allow a much cleaner implementation of downloads tracking

File: bin/travis-build.bash [hunk 1]
```diff
@@ -62,6 +62,11 @@ cd ckanext-hierarchy
 python setup.py develop
 cd -
 
+git clone https://github.com/TkTech/ckanext-cloudstorage
+cd ckanext-cloudstorage
+python setup.py develop
+cd -
```
Comment (chris48s) [inline bin/travis-build.bash:68]: This is a bit annoying. The problem I'm trying to solve here is:

If we try to call `from ckanext.cloudstorage.controller import StorageController` and `ckanext-cloudstorage` isn't installed, we'll throw `ImportError`. In general, this is useful behaviour. If I'm trying to extend `ckanext.cloudstorage.controller.StorageController` and `ckanext-cloudstorage` isn't installed then that's exactly how we want to be told about it - we shouldn't try to fall back to some other behaviour.

This is fine, except that when nose tries to read in everything to measure statement coverage at the end of the test run, trying to parse `extended_storage.py` throws an import error. I think installing `ckanext-cloudstorage` to silence the error but not registering `cloudstorage` in `CKAN__PLUGINS` (so we still use the local storage backend under test - that's what we look at when we set up the routes in `plugin.py`) is probably a reasonable workaround for this.

Another approach would be to explicitly ignore that file for test coverage, but I think it is useful to retain the information that we don't have tests covering this behaviour.

Open to another suggestion though.. I think whatever we do will be a bit of a fiddle

**Comment by chris48s** [comment]
This is ready for review now.

File: bin/travis-build.bash (hunk 1 above)
Comment (amercader) [reply bin/travis-build.bash:68]: Installing and not enabling sounds fine

File: ckanext/unhcr/plugin.py [hunk 2]
```diff
@@ -116,7 +116,7 @@ def before_map(self, _map):
             _map.connect('/dataset/{id}/resource/{resource_id}/download', controller=controller, action='resource_download')
```
Comment (amercader) [inline ckanext/unhcr/plugin.py:116]: Can we combine the logic for local and cloudstorage downloads in a single endpoint? We can have a single `resource_download` function and then do the check to see if cloudstorage is loaded there. 
The existing check (`if 'cloudstorage' not in os.environ.get('CKAN__PLUGINS', '')` is not reliable, is best to do `'cloudstorage' in config['ckan.plugins']`

Comment (chris48s) [reply ckanext/unhcr/plugin.py:117]: I've made the change you've suggested and submitted it as a different PR in #297

Having done that implementation too, I actually think the approach in this PR is slightly more "correct" due to https://github.com/okfn/ckanext-unhcr/pull/297#discussion_r424989752 . What do you think?

Comment (amercader) [reply ckanext/unhcr/plugin.py:117]: It's not a big deal and both approaches have drawbacks so let's go with the PR approach.
I was thinking something along these lines for the other approach:

```python

class ExtendedStorage(CorePackageController):

    def resource_download():

         if cloudstorage:
               return CloudStorageController().resource_download()

        return super(ExtendedStorage).resource_download()

```

But let's keep the two files like in this PR

**Comment by chris48s** [comment]
OK, cool. I've rebased this to resolve the merge conflicts. I will have a look at getting this deployed to the UAT instance on AWS today or Monday so we start logging the downloads there.
I'll also rebase #294 so that can be reviewed.

**Pull request merged**